### PR TITLE
Compute, compare and understand Benchmarks

### PR DESCRIPTION
**Pull request by afiune**
# Automate Benchmarks

 We use benchmarks to discover what is the best performance being achieved, whether in
a particular function, or an entire component/service. This information can then be
used to identify gaps or deltas between code changes and will be the grounds for us
to be able to make decisions in the near and long future.

 ## Components with Benchmarks

 The list of components that has benchmarks:

 - [Applications Service](https://github.com/chef/automate/tree/master/benchmarks/applications-service)

 ## Directory Scructure

 Every component should have its own directory where we will store historical benchmarks
for every pull request that modifies the component:

 ```
   benchmarks/
    |-README.md
    |-applications-service/
    |  |-txt/
    |  |-html/
    |  |-README.md
    |-ingest-service/
    |  |-txt/
    |  |-html/
    |  |-README.md
```

 At the moment we will run the benchmarks manually on our local computers, on every
component directory should have a `README.md` that explains how to run the benchmarks
and how to store them. In the future we would like our CI system to run and store them
automatically.

 ## Comparing Benchmarks (benchstat)

 To compare two different benchmarks we use the go tool `benchstat`, use the last generated
benchmark and the new one from your pull request you are comparing to.

 Example comparing benchmarks for the `applications-service`:

 ```
[1][default:/src:0]# cd /src/benchmarks/applications-service/txt
[2][default:/src/benchmarks/applications-service/txt:0]# install_benchstat
[3][default:/src/benchmarks/applications-service/txt:0]# benchstat PR_OLD.txt PR_NEW.txt
```

## How to run applications-service benchmarks

 - Enter the studio.
```
$ hab studio enter
```
- Build and start the applications-service.
```
[1][default:/src:0]# start_application_service
```
- Run the benchmarks and store them. _(NOTE: You should already have a PR open)_
```
[2][default:/src:0]# applications_benchmarks
  hab-studio: Running applications-service benchmark. (temporarily stored at /tmp/new.txt)
  hab-studio: This might take a while...
  hab-studio: Saving benchmark results at /src/benchmarks/applications-service/txt/new.txt
  hab-studio: Using benchstat tool to compute the results. ('benchstat new.txt')
  hab-studio: Generating HTML of the results at /src/benchmarks/applications-service/html/new.html
  hab-studio: Renaming the generated benchmark files:
               TXT:  /src/benchmarks/applications-service/txt/new.txt
              HTML: /src/benchmarks/applications-service/html/new.html
   hab-studio: Enter the new name based on the pull request opened.
               Example: For the following PR https://github.com/chef/automate/pull/510
                       The suggested name would be 'PR_510'
               New name: [ENTER_NAME_HERE]
  hab-studio: New benchmark files:
               TXT:  /src/benchmarks/applications-service/txt/[NAME_ENTERED].txt
              HTML: /src/benchmarks/applications-service/html/[NAME_ENTERED].html
```

### CODE REVIEW

File: .studio/applications-service [hunk 1]
```diff
@@ -66,6 +66,46 @@ function applications_get_service_groups_unknown() {
...
+    tmp_txt=/tmp/new.txt
+    results_txt=/src/benchmarks/applications-service/txt/new.txt
+    results_html=/src/benchmarks/applications-service/html/new.html
+
+    log_line "Running applications-service benchmark. (temporarily stored at $(yellow $tmp_txt))"
+    log_line "$(yellow "This might take a while...")"
+    cd components/applications-service || return 1
+    A2_SVC_NAME="applications-service" A2_SVC_PATH="/hab/svc/applications-service" go test \
+      ./... \
+      -bench=Benchmark \
```
Comment (srenatus) [inline .studio/applications-service]: 🤔 What about adding `-benchmem`?

File: .studio/applications-service [hunk 2]
```diff
@@ -66,6 +66,87 @@ function applications_get_service_groups_unknown() {
...
+    rename_benchmark_files
+  )
+}
+
+# This function assumes you have a new.txt and new.html files generated
+# @afiune maybe we can make this function flexible for other components
+function rename_benchmark_files() {
+  local results_txt=/src/benchmarks/applications-service/txt/new.txt
+  local results_html=/src/benchmarks/applications-service/html/new.html
+  local name=
```
Comment (srenatus) [inline .studio/applications-service]: [nit] it's the same as 
```suggestion
  local name
```
isn't it?

File: .studio/applications-service [hunk 3]
```diff
@@ -66,6 +66,87 @@ function applications_get_service_groups_unknown() {
...
+    cd components/applications-service || return 1
+    A2_SVC_NAME="applications-service" A2_SVC_PATH="/hab/svc/applications-service" go test \
+      ./... \
+      -bench=Benchmark \
+      -benchmem \
+      -count=5 2>&1 | tee $tmp_txt
+
+    log_line "Saving benchmark results at $(green $results_txt)"
+    grep Benchmark $tmp_txt > $results_txt
+    rm $tmp_txt
```
Comment (srenatus) [inline .studio/applications-service]: could have just done
```
| tee $results_txt
```
above, I think. benchstat will ignore the non-`Benchmark` lines

Comment (afiune) [reply .studio/applications-service]: Ill test this!!! That would be super awesome!!

File: .studio/applications-service (hunk 2 above)
Comment (afiune) [reply .studio/applications-service]: lol! yeah 😅 thanks Stephan

File: benchmarks/applications-service/README.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+## How to run benchmarks
+
+- Enter the studio.
+```
+$ hab studio enter
+```
+- Build and start the applications-service.
+```
+[1][default:/src:0]# start_application_service
```
Comment (lancewf) [inline benchmarks/applications-service/README.md]: This should be `start_applications_service`

Comment (afiune) [reply benchmarks/applications-service/README.md]: Yup! I'll accept a suggestion from you, or I'll go just edit it manually. Thanks though for the review!

Comment (lancewf) [reply benchmarks/applications-service/README.md]:
```suggestion
[1][default:/src:0]# start_applications_service
```

**Pull request merged**